cpp/src/arrow/compute/row/grouper.cc

```diff
@@ -170,42 +169,137 @@ struct SimpleKeySegmenter : public BaseRowSegmenter {
     }
-    return GetNextSegment(*array.type, GetValuesAsBytes(array), offset, batch.length);
+    return GetNextSegmentDeprecated(*array.type, GetValuesAsBytes(array), offset,
+                                    batch.length);
+    ARROW_UNSUPPRESS_DEPRECATION_WARNING
+  }
+
+  Result<std::vector<Segment>> GetSegments(const ExecSpan& batch) override {
+    RETURN_NOT_OK(CheckForGetSegments(batch, {key_type_}));
+
+    if (batch.length == 0) {
+      return std::vector<Segment>{};
+    }
+
+    const auto& value = batch.values[0];
+    RETURN_NOT_OK(CheckType(*value.type()));
+
+    std::vector<Segment> segments;
+    const void* key_data;
+    if (value.is_scalar()) {
+      const auto& scalar = *value.scalar;
+      DCHECK(scalar.is_valid);
+      key_data = checked_cast<const PrimitiveScalarBase&>(scalar).data();
+      bool extends = Extend(key_data);
+      segments.push_back(MakeSegment(batch.length, 0, batch.length, extends));
+    } else {
+      DCHECK(value.is_array());
+      const auto& array = value.array;
+      DCHECK_EQ(array.GetNullCount(), 0);
+      auto data = GetValuesAsBytes(array);
+      int64_t byte_width = array.type->byte_width();
+      int64_t offset = 0;
+      bool extends = Extend(data);
+      while (offset < array.length) {
+        int64_t match_length = GetMatchLength(data + offset * byte_width, byte_width,
+                                              data, offset, array.length);
+        segments.push_back(MakeSegment(array.length, offset, match_length,
+                                       offset == 0 ? extends : false));
+        offset += match_length;
+      }
+      key_data = data + (array.length - 1) * byte_width;
+    }
+
+    SaveKeyData(key_data);
+
+    return segments;
+  }
+
+ private:
+  static Status CheckType(const DataType& type) {
+    if (!is_fixed_width(type)) {
+      return Status::Invalid("SimpleKeySegmenter does not support type ", type);
+    }
+    return Status::OK();
+  }
+
+  static const uint8_t* GetValuesAsBytes(const ArraySpan& data, int64_t offset = 0) {
+    DCHECK_GT(data.type->byte_width(), 0);
+    int64_t absolute_byte_offset = (data.offset + offset) * data.type->byte_width();
+    return data.GetValues<uint8_t>(1, absolute_byte_offset);
+  }
+
+  // Find the match-length of a value within a fixed-width buffer
+  static int64_t GetMatchLength(const uint8_t* match_bytes, int64_t match_width,
+                                const uint8_t* array_bytes, int64_t offset,
+                                int64_t length) {
+    int64_t cursor, byte_cursor;
+    for (cursor = offset, byte_cursor = match_width * cursor; cursor < length;
+         cursor++, byte_cursor += match_width) {
+      if (memcmp(match_bytes, array_bytes + byte_cursor,
+                 static_cast<size_t>(match_width)) != 0) {
+        break;
+      }
+    }
+    return std::min(cursor, length) - offset;
+  }
+
+  bool Extend(const void* data) {
+    if ARROW_PREDICT_FALSE (!extend_was_called_) {
```

Comment:
Please let's put parentheses here
```suggestion
    if (ARROW_PREDICT_FALSE(!extend_was_called_)) {
```